cpp/include/raft/distance/distance.cuh

```diff
@@ -252,71 +250,63 @@ void pairwise_distance(raft::resources const& handle,
                        bool isRowMajor = true,
                        Type metric_arg = 2.0f)
 {
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  auto dispatch = [&](auto distance_type) {
+    auto worksize = getWorkspaceSize<distance_type(), Type, Type, Type, Index_>(x, y, m, n, k);
+    workspace.resize(worksize, stream);
+    detail::distance<distance_type(), Type, Type, Type, Index_>(
```

Comment:
I guess we still need to know some things ahead of time for loop unrolling. I'm wondering if there's still any way we could capture (and consolidate) those differences in templates so we can eventually get to the point where the distances themselves no longer need to be separate templates?